src/inbox.rs

```diff
@@ -148,13 +149,7 @@ impl<A> Chan<A> {
                 Ok(inner.pop_priority().unwrap().into())
             }
             // Shared priority is less - take from broadcast
-            Ordering::Less => {
-                let msg = broadcast.pop().unwrap().0;
-                drop(broadcast);
-                inner.try_advance_broadcast_tail();
-
-                Ok(msg.into())
-            }
+            Ordering::Less => Ok(inner.pop_broadcast(broadcast_mailbox).unwrap().into()),
```

Comment:
I'm not sure we can do this, since the broadcast priority could be out of date, as the channel could get modified between when we check it and when we take from it

Reply:
Ah, that makes sense. I can lock it at the top now though I think. That wasn't possible before.

Reply:
Damn, doesn't work actually because we lock it again inside.

Reply:
If we move the locking of `ChanInner` further up, I don't think there can be a race condition though. We only send into the broadcast mailbox with `ChanInner` locked.

Reply:
Yea, but wouldn't another receive future on the same mailbox be able to take out a message? Concurrent sends are not a problem since they can never decrease the max element of the heap.

Reply:
I don't think so. The `ReceiveFuture` just calls `try_recv` which locks `ChanInner` first thing. How would a concurrent receive be possible?

Reply:
Sorry for getting back to this late - I think I was under the impression that we still only supplied the waiting rx with a notification to check its mailbox rather than the message itself